### PR TITLE
Add tunterm acl plugin support.

### PR DESCRIPTION
**Pull request by dypet**
### Overview:
These changes add support for the recently added tunterm_acl vpp plugin (https://github.com/sonic-net/sonic-platform-vpp/pull/114) in sonic-vpp. 

### Changes:
Added API calls to the new plugin in SaiVppXlate.c, and added logic to SwitchStateBaseAcl.cpp so that the new plugin is called with the appropriate fields instead of the regular ACL plugin when SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is received. Ports are recorded in m_acl_tbl_hw_ports_map at table creation so that they may be passed to the tunterm_acl plugin and bound by the new plugin the first time an ACE with SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is configured. tunterm_acl ACLs are tracked in m_tunterm_acl_swindex_map similarly to m_acl_swindex_map for a given tbl_oid.

### CODE REVIEW

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (abdbaig) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:904]: indent this block

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (abdbaig) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1646]: remove additional lines

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c
Comment (abdbaig) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:1981]: Here too you can use vpp_to_vl_api_ip_addr() to convert IP addresses

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:745]: Fix indentation here and throughout.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1627]: In the unbind case, wouldn't hwif_list always be empty?

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1642]: Does this correctly update the ports_map in the unbind case?

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:837]: Collapse these into a single if statement as they are part of the same check.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:843]: Where possible, follow formatting convention of the file (eg. add a space after your if)

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.h
Comment (AkeelAli) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.h:96]: Correct structure name to vpp_tunterm_acl_t

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:901]: Collapse into 1 if as it's part of the same check.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:938]: This block shouldn't have changed, yet it's marked as a diff here due to indentation changes. Please revert this.

**Review by AkeelAli — COMMENTED** [review]
If not already done, can you please re-run the basic ACL redirect tests with the latest plugin (as some changes were made before it was merged). You can wait until the final version of this review to do that.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:830]: This function is too long. We'd better break it into a few smaller functions. For example, for the part that converting ACE to acl_tbl_entries_t and sort can be one function. From sorted  acl_tbl_entries_t to vpp_acl_t and vpp_tunerm_acl_t can be another function.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:924]: Can you explain this?

File: platform/saivpp/vpplib/SwitchStateBase.h
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBase.h:1021]: Please add some documents to explain these new functions

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:987]: This should be an error

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:471]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.c]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Move the ports_map unbind update above directly in the unbind case

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: In this case want to still break out of the loop when SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is false - updated the same check later on to be in a single 'if' statement.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.h
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.h]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: The check for acl->count > 0 was added here to skip this block for the cases where there are only tunterm acl rules and no standard acl rules

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: When iterating through the incoming rule's sai attributes we don't know if it's a tunterm acl or regular acl until SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is encountered. There are common attr between the two, like SAI_ACL_ENTRY_ATTR_FIELD_DST_IP, which would need to be populated in both cases. Both rules (regular acl and tunterm acl) are updated in parallel while looping through the sai attributes, and then at the end once we know if it's a tunterm rule or not we clear the other rule which potentially had fields populated.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (yue-fred-gao) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:798]: Conceptually with tunterm feature, each ACL table can have 1 or 2 ACL objects in VPP: one for regular ACL and one for tunterm ACL. In the create/update request, we go through the ACEs and convert them to 2 lists of vpp_acl_t and vpp_tunterm_acl_t (you can create a super data structure including the 2 lists). Then we pass the 2 lists along with table_id to a function. The function reads acl_swindex and tunterm_acl_swindex from table_id and does necessary operation to create/update/delete acl or tunterm acl from vpp.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:522]: Please change all NOTICE to INFO except very few important one like tunterm acl created/modify/deleted successfully. See issue #116 .

File: platform/saivpp/vpplib/SwitchStateBase.h
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBase.h:995]: There is already m_acl_tbl_grp_ports_map. If we see acl and tunterm_acl as two objects created from the same acl_table, it is possible to share the same mapping. even m_acl_swindex_map and m_tunterm_acl_swindex_map can be combined to one if we use std::pair or std::Array to keep both indexes in one

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:2119]: why do we need it in context? Is it already passed in with tunterm_acl_index?

File: platform/saivpp/vpplib/SwitchStateBase.h
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBase.h]: Added

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1188]: Regular ACL will still exist with default config and will need to be cleaned up

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBase.h
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBase.h:996]: Will consider merging some of these mappings as a future enhancement

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:2119]: In create case the VPP tunterm sw_index will be returned here

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:928]: Just return the call. It makes it clear we are short-circuiting this function call here.

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:911]: Any reason you don't check status of `vpp_tunterm_acl_interface_add_del` or `tunterm_acl_bindunbind`?

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:947]: Optional: add a comment here explaining the intent of this block, ie. "enable tunterm ACL on ports to which the table is already bound"

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:1000]: You've already checked STATUS_SUCCESS above, you can move this check on table_delete inside the same block or return on non success.

**Review by AkeelAli — APPROVED** [review]
Minor comments. We can add an HLD later to detail the feature + how to test it.

File: saivpp/src/vppxlate/SaiVppXlate.h
Comment (dypet) [reply saivpp/src/vppxlate/SaiVppXlate.h]: Updated

File: saivpp/src/vppxlate/SaiVppXlate.h
Comment (dypet) [reply saivpp/src/vppxlate/SaiVppXlate.h]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp:926]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp:958]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

**Pull request merged**
